main/src/library/Iterator.flix

```diff
@@ -216,12 +262,72 @@ namespace Iterator {
     ///
     /// Returns the longest prefix of `iter` that satisfies the predicate `f`.
     ///
-    /// The function `f` must be pure.
+    /// Whether `f` is applied eagerly or lazily depends on its purity:
+    /// - If `f` is pure then it is applied lazily.
+    /// - If `f` is impure then it is applied eagerly (consuming elements of the iterator until and including
+    /// the first element that does *not* satisfy predicate `f`).
     ///
-    /// Consumes elements of the iterator until the first element that does *not* satisfy predicate `f`.
+    /// The original iterator `iter` should not be reused.
+    ///
+    pub def takeWhile(f: a -> Bool & ef, iter: Iterator[a]): Iterator[a] & Impure =
+        matchEff f {
+            case Pure   => takeWhileL(f as a -> Bool, iter)
+            case Impure => takeWhileE(f as a -> Bool & Impure, iter)
+        }
+
+    ///
+    /// Returns the longest prefix of `iter` that satisfies the predicate `f`.
+    /// `f` is applied *lazily*.
+    ///
+    /// Does *not* consume any elements of the iterator.
+    ///
+    /// The original iterator `iter` should not be reused.
+    ///
+    def takeWhileL(f: a -> Bool, iter: Iterator[a]): Iterator[a] & Impure =
+        let Iterator(done, next) = iter;
+        if (done())
+            Iterator(() -> true as & Impure, () -> bug!("Empty iterator") as & Impure)
```

Comment:
Is there a reason we can't just return the same iterator?

Reply:
I agree we should return the same iterator, but I saw Magnus using this and wanted it to be consistent with that. I think I've returned `iter` other places so it's not consistent across the board.

Reply:
Maybe we can summon @magnus-madsen to see if there's a specific reason for it.

Reply:
I think we should return the original iterator whenever we can, if I have not done so somewhere that was unintentional.